### PR TITLE
Sparse Group Lasso

### PR DESCRIPTION
**Pull request by wassimmazouz**
Adding Sparse Group Lasso compatibility

### CODE REVIEW

File: solvers/gsroptim.py [hunk 1]
```diff
@@ -45,3 +45,7 @@ def run(self, n_iter):
 
     def get_result(self):
         return dict(beta=self.coef)
+
```
Comment (wassimmazouz) [inline solvers/gsroptim.py]: @mathurinm is warmup needed for `gsroptim` too?

**Comment by mathurinm** [comment]
No, they use cython so their code is already compiled once and for all at
install time

Il mer 31 lug 2024, 15:15 Wassim MAZOUZ ***@***.***> ha
scritto:

File: README.rst [hunk 2]
```diff
@@ -1,20 +1,20 @@
-Benchmark Group Lasso
-=====================
+Benchmark Sparse Group Lasso
```
Comment (mathurinm) [inline README.rst]:
```suggestion
Sparse Group Lasso Benchmark
```

File: README.rst [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
+============================
 |Build Status| |Python 3.6+|
 
-This benchmark is dedicated to solver of **Group Lasso problem**. The problem reads
+This benchmark is dedicated to solver of **Sparse Group Lasso problem**. The problem reads
```
Comment (mathurinm) [inline README.rst]:
```suggestion
This benchmark is dedicated to the **Sparse Group Lasso**. The optimization problem reads
```

File: example_config.yml [hunk 4]
```diff
@@ -4,7 +4,7 @@ dataset:
   - Drugs Potency
 
 objective:
-  - Group Lasso objective[reg=[1,1e-1,1e-2,1e-3]]
+  - Sparse Group Lasso objective[reg=[1,1e-1,1e-2,1e-3]]
```
Comment (mathurinm) [inline example_config.yml]: why objective? remove it to be shorter

File: objective.py [hunk 5]
```diff
@@ -5,70 +5,86 @@
     import numpy as np
     from numpy.linalg import norm
 
-    from skglm.datafits import QuadraticGroup
-    from skglm.penalties import WeightedGroupL2
     from skglm.utils.data import grp_converter
 
 
 class Objective(BaseObjective):
-    name = "Group Lasso objective"
```
Comment (mathurinm) [inline objective.py:14]: remove objective

File: objective.py [hunk 6]
```diff
@@ -5,70 +5,86 @@
...
-
-    requirements = ["pip:skglm"]
-
+    name = "Sparse Group Lasso objective"
+
+    requirements = [
+        "numpy'<2'",
+        "pip::git+https://github.com/scikit-learn-contrib/skglm",
+        "pip::git+https://github.com/EugeneNdiaye/Gap_Safe_Rules"
```
Comment (mathurinm) [inline objective.py]: add a comma at the end of the line

File: objective.py [hunk 7]
```diff
@@ -5,70 +5,86 @@
...
 
 class Objective(BaseObjective):
-    name = "Group Lasso objective"
-
-    requirements = ["pip:skglm"]
-
+    name = "Sparse Group Lasso objective"
+
+    requirements = [
+        "numpy'<2'",
```
Comment (mathurinm) [inline objective.py]: explain why each package is needed for the whole benchmark (not just for a single solver)

File: solvers/gsroptim.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+with safe_import_context() as import_ctx:
+    import numpy as np
+    from gsroptim.sgl import sgl_path
+
+
+class Solver(BaseSolver):
+    """Eugene Ndiaye, Olivier Fercoq, Alexandre Gramfort, and Joseph Salmon.
+    2016. GAP safe screening rules for Sparse-Group Lasso. In Proceedings
+    of the 30th International Conference on Neural Information Processing
+    Systems (NIPS'16). Curran Associates Inc., Red Hook, NY, USA, 388–396.
```
Comment (mathurinm) [inline solvers/gsroptim.py]: Be shorter: NeurIPS 2016 (not NIPS)

**Review by mathurinm — COMMENTED** [review]
Looks clean, thanks @wassimmazouz ! Can you test it on a small config to check that all solvers go to the same solution? (no issue with scaling of lambdas, etc)

**Comment by wassimmazouz** [comment]
@mathurinm all solvers seem to behave correctly in [all possible configs](https://wassimmazouz.github.io/benchmarks/benchmark_sparse_group_lasso.html).

File: objective.py [hunk 9]
```diff
@@ -5,70 +5,86 @@
...
-                norm(X[:, grp_g_indices].T @ y) / n_samples
+            grp_g_indices = self.grp_indices[
+                self.grp_ptr[g]: self.grp_ptr[g+1]]
+            lmbd_max_group = max(
+                lmbd_max_group,
+                norm(self.X[:, grp_g_indices].T @ self.y) / self.n_samples
             )
 
+        # Combine the two parts using the tau parameter
+        lmbd_max = self.tau * lmbd_max_l1 + (1 - self.tau) * lmbd_max_group
```
Comment (mathurinm) [inline objective.py:87]: I don't think this is correct. The dual norm is the epsilon norm, which is harder than this to compute (that's why we needed gsroptim). 
Test at reg=0.99 and reg=1.01 but I don't think the current formula is the good one.
@Badr-MOUFAD

**Comment by wassimmazouz** [comment]
@Badr-MOUFAD could you take a look at this? Thanks

**Review by mathurinm — CHANGES_REQUESTED** [review]
check lambda_max

File: objective.py (hunk 9 above)
Comment (Badr-MOUFAD) [reply objective.py:87]: Yes indeed, 
as the lambda max is hard to get in this case, I opted for an upper bound which happen to be cheap to get

File: objective.py [hunk 10]
```diff
@@ -5,70 +5,86 @@
...
 
 
 class Objective(BaseObjective):
-    name = "Group Lasso objective"
-
-    requirements = ["pip:skglm"]
-
+    name = "Sparse Group Lasso"
+
+# skglm is needed here to create group partition and indices
```
Comment (Badr-MOUFAD) [inline objective.py]:
```suggestion
    # skglm is needed here to create group partition and indices
```

File: objective.py [hunk 11]
```diff
@@ -5,70 +5,86 @@
...
             X=self.X, y=self.y, lmbd=self.lmbd, groups=self.groups,
-            grp_indices=self.grp_indices, grp_ptr=self.grp_ptr
+            grp_indices=self.grp_indices, grp_ptr=self.grp_ptr, tau=self.tau
         )
 
-    def _compute_lmbd_max(self, X, y, grp_indices, grp_ptr):
-        lmbd_max = 0.
-        n_samples = X.shape[0]
-        n_groups = len(grp_ptr) - 1
+    def _compute_lmbd_max(self):
```
Comment (Badr-MOUFAD) [inline objective.py:69]: As pointed out by @mathurinm in https://github.com/Badr-MOUFAD/benchmark_group_lasso/pull/1/files#r1701936766,
this is a heuristic to compute the lambda max (when ``tau == 0``, it is provides the correct `lmbd_max`)

We better add a comment to mention that

**Review by Badr-MOUFAD — COMMENTED** [review]
Thanks @wassimmazouz for the hard work 💪 

Except for some minor changes to be made, LGTM 🚀

File: example_config.yml [hunk 12]
```diff
@@ -8,4 +8,3 @@ objective:
 
 max-runs: 50
 n-repetitions: 1
-n-jobs: 15
```
Comment (wassimmazouz) [inline example_config.yml]: I removed this as it seems to limit `skglm` to one iteration

Comment (mathurinm) [reply example_config.yml]: I don't think this is related, can you isolate this

Comment (mathurinm) [reply example_config.yml]: @wassimmazouz

File: objective.py [hunk 13]
```diff
@@ -5,70 +5,89 @@
...
-    requirements = ["pip:skglm"]
-
+    name = "Sparse Group Lasso"
+
+    # skglm is needed here to create group partition and indices
+    requirements = [
+        "numpy'<2'",
+        "pip::git+https://github.com/scikit-learn-contrib/skglm",
+    ]
+    # tau = 0 is Group Lasso
```
Comment (mathurinm) [inline objective.py]:
```suggestion
    # tau = 0 is Group Lasso, tau = 1 is Lasso
```

File: objective.py [hunk 14]
```diff
@@ -5,70 +5,89 @@
...
 class Objective(BaseObjective):
-    name = "Group Lasso objective"
-
-    requirements = ["pip:skglm"]
-
+    name = "Sparse Group Lasso"
+
+    # skglm is needed here to create group partition and indices
+    requirements = [
+        "numpy'<2'",
```
Comment (mathurinm) [inline objective.py]: why is numpy <2 needed ?

File: objective.py [hunk 15]
```diff
@@ -5,70 +5,89 @@
...
 
+        # Combine the two parts using the tau parameter
+        lmbd_max = self.tau * lmbd_max_l1 + (1 - self.tau) * lmbd_max_group
+
         return lmbd_max
 
     def get_one_result(self):
-        n_features = self.X.shape[1]
-        return dict(w=np.zeros(n_features))
+        return dict(w=np.zeros(self.n_features))
```
Comment (mathurinm) [inline objective.py]: the rest of the code uses beta

File: solvers/skglm.py [hunk 16]
```diff
@@ -39,7 +51,7 @@ def run(self, n_iter):
         )
 
     def get_result(self):
-        return dict(w=self.w)
+        return dict(beta=self.w)
```
Comment (mathurinm) [inline solvers/skglm.py]: define self.beta instead of self.w for harmonization (also true for other solvers); this will be less confusing

**Review by mathurinm — APPROVED** [review]

**Pull request merged**